rust/gitxetcore/src/git_integration/git_repo_plumbing.rs

```diff
@@ -357,14 +359,148 @@ pub fn list_files_from_repo(
     Ok(list)
 }
 
+/// Walk the repo working directory starting from search_root.
+/// Return a list of file paths under the search_root, the
+/// file paths are relative to the working dir root.
+/// Note that symlinks are ignored because they are difficult to
+/// deal with: git deals with the symlink file itself without
+/// following the link.
+pub fn walk_working_dir(
+    work_root: impl AsRef<Path>,
+    search_root: impl AsRef<Path>,
+    recursive: bool,
+) -> anyhow::Result<Vec<PathBuf>> {
+    // check existence
+    let work_root = work_root.as_ref();
+    let search_root = search_root.as_ref();
+
+    if !work_root.exists() || !search_root.exists() {
+        return Ok(vec![]);
+    }
+
+    // get absolute paths and check if search below work root
+    let work_root = work_root.canonicalize()?;
+    let search_root = search_root.canonicalize()?;
+
+    if !search_root.starts_with(&work_root) {
+        return Err(anyhow!(
+            "Path {search_root:?} is outside repository at {work_root:?}"
+        ));
+    }
+
+    // ignore .git folder, .gitignore, .gitattributes
+    if is_git_special_files(
+        search_root
+            .file_name()
+            .unwrap_or_default()
+            .to_str()
+            .unwrap_or_default(),
+    ) {
+        return Ok(vec![]);
+    }
+
+    // directly return single file itself
+    if search_root.is_file() {
+        return Ok(vec![search_root.strip_prefix(work_root)?.to_owned()]);
+    }
+
+    // now search_root is a regular directory, walk below it
+
+    let mut ret = vec![];
+
+    let mut queue = VecDeque::from([search_root]);
+
+    while !queue.is_empty() {
+        let sr = queue.pop_front().unwrap();
+
+        let dir_walker = std::fs::read_dir(sr)?;
+
+        for entry in dir_walker.flatten() {
+            let file_type = entry.file_type()?;
+            let file_path = entry.path();
+
+            if file_type.is_file() {
+                ret.push(file_path.strip_prefix(&work_root)?.to_owned());
+            } else if recursive
```

Comment:
I think we need a check here for is_dir(); if it's a symlink, it ends up getting put on the queue and treated as a directory.

Reply:
Good point. I've rewritten this part using walkdir, should exclude symlinks.